awswrangler/dynamodb/_read.py

```diff
@@ -146,3 +150,331 @@ def read_partiql_query(
     _logger.debug("Reading results for PartiQL query:  %s", query)
     items = _get_items(query=query, boto3_session=boto3_session)
     return _parse_dynamodb_items(items=items, dtype=dtype)
+
+
+def _get_invalid_kwarg(msg: str) -> Optional[str]:
+    """Detect which kwarg contains reserved keywords based on given error message.
+
+    Parameters
+    ----------
+    msg : str
+        Botocore client error message.
+
+    Returns
+    -------
+    str, optional
+        Detected invalid kwarg if any, None otherwise.
+    """
+    for kwarg in ("ProjectionExpression", "KeyConditionExpression", "FilterExpression"):
+        if msg.startswith(f"Invalid {kwarg}: Attribute name is a reserved keyword; reserved keyword: "):
+            return kwarg
+    return None
+
+
+# SEE: https://stackoverflow.com/a/72295070
+CustomCallable = TypeVar("CustomCallable", bound=Callable[[Any], Sequence[Dict[str, Any]]])
+
+
+def _handle_reserved_keyword_error(func: CustomCallable) -> CustomCallable:
+    """Handle automatic replacement of DynamoDB reserved keywords.
+
+    For reserved keywords reference:
+    https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ReservedWords.html.
+    """
+
+    @wraps(func)
+    def wrapper(*args: Any, **kwargs: Any) -> Sequence[Dict[str, Any]]:
+        try:
+            return func(*args, **kwargs)
+        except ClientError as e:
+            error_code, error_message = (e.response["Error"]["Code"], e.response["Error"]["Message"])
+            # Check catched error to verify its message
+            kwarg = _get_invalid_kwarg(error_message)
+            if (error_code == "ValidationException") and kwarg:
+                reserved_keyword = error_message.split("keyword: ")[-1]
+                sanitized_keyword = f"#{reserved_keyword}"
+                kwargs[kwarg] = kwargs[kwarg].replace(reserved_keyword, sanitized_keyword)
+                kwargs["ExpressionAttributeNames"] = {
+                    **kwargs.get("ExpressionAttributeNames", {}),
+                    sanitized_keyword: reserved_keyword,
+                }
+                # SEE: recursive approach guarantees that each reserved keyword will be properly replaced,
+                # even if it will require as many calls as the reserved keywords involved (not so efficient...)
+                return wrapper(*args, **kwargs)
```

Comment:
@jaidisido do you have any feedback on this?

I wasn't able to find any way to obtain the list of DynamoDB reserved keywords programmatically (having that beforehand would ease the implementation without requiring recursion).

Reply:
I did look/ask around and there is no programmatic way to access the list of reserved keywords sadly.

That being said, looking at its [GH history](https://github.com/awsdocs/amazon-dynamodb-developer-guide/blob/master/doc_source/ReservedWords.md) that list is static. It has not been changed since its inception.

This makes sense as adding a keyword to the list would be a serious breaking change in production.

All of this to say that I would be ok with creating a separate file hosting the list of keywords and using it for your check ([example](https://github.com/aws/aws-sdk-js-v3/blob/140b9b339059f833aa2d68a9d4833422282dc229/packages/util-dynamodb/src/isReservedWord.ts)).

Reply:
Thanks for the reference. 

I tried to switch to a LBYL approach for reserved words management but I found it harder than expected - mainly due to the diverse format which `ProjectionExpression`, `KeyConditionExpression` and `FilterExpression` can assume - making difficult to find a general preventive action.

So, given that:

- expressions can be also expressed via `Key` and `Attr`, which seem to sanitize reserved words too
- the (current) EAFP approach is already available

I would propose to keep it as it is, at least for the moment.

If we are worried about the recursive approach (which leads to as many validation client calls in case as number of reserved words in the given kwargs), we can put it behind an opt-in/opt-out kwargs, something like `sanitize_dynamodb_reserved_words: bool`.

Reply:
Agreed, I think it's ok to keep it as is for now as it would be too complex to capture all edge cases from the involved expressions